src/Log/Formatter/AbstractFormatter.php

```diff
@@ -0,0 +1,42 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) :  Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakefoundation.org CakePHP(tm) Project
+ * @since         4.3.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Log\Formatter;
+
+use Cake\Core\InstanceConfigTrait;
+
+abstract class AbstractFormatter
```

Comment:
You could also have an interface and have each class use the `InstanceConfigTrait` directly (if they want to).

Reply:
I was using an Interface, but I ended up with warnings because the ConsoleLog, FileLog and SyslogLog constructors were calling `setConfig()` on an interface.

Reply:
Oh. you'd have to have `setConfig()` in the interface as well if the logging engines are calling it. You could avoid having `setConfig()` in the interface if you shim the options differently but it would require slightly more complex shim code as you'd need to handle `formatter` being a string and an array.

You could do something like

```php
// in the logging engine
 if (isset($this->_config['format'])) {
    deprecationWarning('`format` option should now be set in the formatter options.');
    // new protected method to modify the formatter config.
    $config = $this->setFormatterOption($config, 'format');
}
parent::__construct($config);
```

Reply:
Typically, you use an Interface when passing an instance in. In this case, we're always instantiating the formatter so I'm not sure how well that works. We need it to have a constructor that accepts a $config array as well.